jaeger/cmd/check.go

```diff
@@ -0,0 +1,257 @@
+	}
+	return nil
+}
+
+func newCmdCheck() *cobra.Command {
+	options := newCheckOptions()
+	cmd := &cobra.Command{
+		Use:   "check [flags]",
+		Args:  cobra.NoArgs,
+		Short: "Check the Jaeger extension for potential problems",
+		Long: `Check the jaeger extension for potential problems.
+
+The check command will perform a series of checks to validate that the Jaeger extension
+ is configured correctly. If the command encounters a
+failure it will print additional information about the failure and exit with a
+non-zero exit code.`,
+		Example: `  # Check that the Jaeger extension is up and running
+  linkerd jaeger check`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			// Get jaeger Extension Namespace
+			ns, err := getNamespaceOfExtension("linkerd-jaeger")
+			if err != nil {
+				fmt.Fprint(os.Stderr, err.Error())
+				os.Exit(1)
+			}
+			jaegerNamespace = ns.Name
+
+			return configureAndRunChecks(stdout, stderr, options)
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.output, "output", "o", options.output, "Output format. One of: basic, json")
+	cmd.PersistentFlags().DurationVar(&options.wait, "wait", options.wait, "Maximum allowed time for all tests to pass")
+
+	return cmd
+}
+
+func configureAndRunChecks(wout io.Writer, werr io.Writer, options *checkOptions) error {
+	err := options.validate()
+	if err != nil {
+		return fmt.Errorf("Validation error when executing check command: %v", err)
+	}
+
+	checks := []healthcheck.CategoryID{
+		linkerdJaegerExtensionCheck,
+	}
+
+	hc := healthcheck.NewHealthChecker(checks, &healthcheck.Options{
+		ControlPlaneNamespace: controlPlaneNamespace,
+		KubeConfig:            kubeconfigPath,
+		KubeContext:           kubeContext,
+		Impersonate:           impersonate,
+		ImpersonateGroup:      impersonateGroup,
+		APIAddr:               apiAddr,
+		RetryDeadline:         time.Now().Add(options.wait),
+		MultiCluster:          false,
+	})
+
+	hc.AppendCategories(*jaegerCategory())
+
+	success := healthcheck.RunChecks(wout, werr, hc, options.output)
+
+	if !success {
+		os.Exit(1)
+	}
+
+	return nil
+}
+
+func getNamespaceOfExtension(name string) (*corev1.Namespace, error) {
+	kubeAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	namespaces, err := kubeAPI.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{LabelSelector: k8s.LinkerdExtensionLabel})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ns := range namespaces.Items {
+		if ns.Labels[k8s.LinkerdExtensionLabel] == name {
+			return &ns, err
+		}
+	}
+	return nil, fmt.Errorf("Could not find the namespace for extension %s", name)
+}
+
+func checkIfDataPlanePodsExist(pods []corev1.Pod) error {
+	for _, pod := range pods {
+		proxyContainer := false
+		for _, containerSpec := range pod.Spec.Containers {
+			if containerSpec.Name == k8s.ProxyContainerName {
+				proxyContainer = true
```

Comment:
can't we just immediately return nil here?

Reply:
We can do that but this way we can also report which pod does not have the proxy container :thinking: but its limited only for the first pod!

Reply:
oh, I see, this is inside the pods loop.  I think this would be clearer if we had a method that takes a pod and checks if that pod has a proxy container.  If we did it that way then we could immediately return nil from that method when a proxy contain is detected.  With that in place, this method becomes much simpler since we can immediately return an error when we encounter a pod without a proxy container.

Reply:
But we would be still needing the boolean flag to check for the proxy container right? I updated the fun to be different ones, as you suggested!